### PR TITLE
Variables: Add setting 'verifyOptionFromUrl' to check option get from URL

### PR DESCRIPTION
**Pull request by jiangpengcheng**
**What this PR does / why we need it**: 

This PR adds a new boolean setting to dashboard variables, and if it's turned on, grafana will check whether option passed via url matches any existing options, if not, reject this option and fallback to the first available option.

I think this will be useful in cases when we want to limit users to access variables which donot belong to them

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

### CODE REVIEW

File: docs/sources/variables/verify-option-get-from-url.md [hunk 1]
```diff
@@ -0,0 +1,8 @@
++++
+title = "Verify option get from url"
+weight = 500
++++
+
+# Verify option
+
+If you turn this on, then when an option is passed via url, and it doesn't match any existing options, grafana will reject this option and fallback to the first available option(or empty value if there are no options at all).
```
Comment (oddlittlebird) [inline docs/sources/variables/verify-option-get-from-url.md:8]:
```suggestion
If you turn this on, then when an option is passed via URL, and it doesn't match any existing options, Grafana rejects this option and fallback to the first available option (or empty value if there are no options at all).
```

File: docs/sources/variables/verify-option-get-from-url.md [hunk 2]
```diff
@@ -0,0 +1,8 @@
++++
+title = "Verify option get from url"
+weight = 500
++++
+
+# Verify option
```
Comment (oddlittlebird) [inline docs/sources/variables/verify-option-get-from-url.md:6]: Where do you turn this on? If it is in the config file, then I recommend updating configuration.md.

Also, I don't understand what this option is or why I might want to use it. Examples or a simple use case would be helpful if this documentation stands.

File: docs/sources/variables/verify-option-get-from-url.md [hunk 3]
```diff
@@ -0,0 +1,8 @@
++++
+title = "Verify option get from url"
```
Comment (oddlittlebird) [inline docs/sources/variables/verify-option-get-from-url.md:2]:
```suggestion
title = "Verify option"
```

Comment (oddlittlebird) [reply docs/sources/variables/verify-option-get-from-url.md:2]: The side menu title should match the h1 title.

File: docs/sources/variables/verify-option-get-from-url.md (hunk 2 above)
Comment (osg-grafana) [reply docs/sources/variables/verify-option-get-from-url.md:6]: When you find out, please update the title of this PR so it makes sense in context of RNs.

**Comment by jiangpengcheng** [comment]
@oddlittlebird 
this options is set in edit page of variable
image

and with this option turned on, if option value get from url doesn't match any exiting options, grafana will use the first existing option:

image

File: docs/sources/variables/verify-option-get-from-url.md (hunk 3 above)
Comment (jiangpengcheng) [reply docs/sources/variables/verify-option-get-from-url.md:2]: updated

File: docs/sources/variables/verify-option-get-from-url.md (hunk 1 above)
Comment (jiangpengcheng) [reply docs/sources/variables/verify-option-get-from-url.md]: thanks for suggestions, updated

**Review by mckn — COMMENTED** [review]
Thanks for opening this PR! I like the added functionality but I think the naming is a bit confusing. I will also include @hugohaggmark as a reviewer to get his input on if this could be default behavior instead.

But if I understood the behavior correctly it is basically something like:

"Allow only values available in options"

**Review by hugohaggmark — COMMENTED** [review]
Hi @jiangpengcheng and thanks for opening this PR, I think you should create an issue first where you describe this feature, the use cases you think it solves and see what the community thinks about this feature before we introduce this behavior as default. 

Also, I think it would confuse users to set this behavior per variable, wouldn't this be better as a setting for the Grafana Org? Regarding security, you should be aware that even viewers can access data sources: https://grafana.com/docs/grafana/latest/administration/security/#limit-viewer-query-permissions

Thank you!

**Comment by jiangpengcheng** [comment]
> Thanks for opening this PR! I like the added functionality but I think the naming is a bit confusing. I will also include @hugohaggmark as a reviewer to get his input on if this could be default behavior instead.
> 
> But if I understood the behavior correctly it is basically something like:
> 
> "Allow only values available in options"

Thanks for correct, I'm not a native english speaker, so may made some mistakes

**Comment by jiangpengcheng** [comment]
@hugohaggmark

Thanks for suggestions, I created an issue here https://github.com/grafana/grafana/issues/35791.
And I made this behavior per variable cause I think only some of variables are sensitive

**Comment by hugohaggmark** [comment]
Thank you @jiangpengcheng, I think the security reason is not a valid use case because a user with the view role can still figure out how to query the data sources directly. But let's keep this open and see if it receives more feedback.

Again thank you for taking the time to contribute!

**Comment by jiangpengcheng** [comment]
> Thank you @jiangpengcheng, I think the security reason is not a valid use case because a user with the view role can still figure out how to query the data sources directly. But let's keep this open and see if it receives more feedback.
> 
> Again thank you for taking the time to contribute!

but I think viewer is not allowed to see how a variable is configured, right?

**Comment by joshhunt** [comment]
Hi @jiangpengcheng! 👋🏻 

Thanks for the contribution, but we're not going to merge this PR into Grafana. While I understand your motivation behind this, I don't think it actually solves a security concern, and might actually give users a false-sense of security around variables being used for sensitive reasons.

A user can still manually issue their own queries by modifying the network requests the Grafana frontend makes. Additionally, a user can view the dashboard JSON to figure out how the variables "work" and potentially circumvent any frontend filters that's been set up.

It's the responsibility of the remote datasource/database (not Grafana or the data source plugin) to ensure that all data is suitable to be viewed by users with a View permission.

If you have a feature request around this for Grafana, [open a discussion](https://github.com/grafana/grafana/discussions) so we can talk about it.

Again - thank you!

**Pull request closed without merging**